### PR TITLE
Support GCS for delta sharing Server

### PR DESCRIPTION
**Pull request by kohei-tosshy**
This is a PR for #20.

I implemented GCS support by using Google Cloud Client Library (https://cloud.google.com/storage/docs/reference/libraries) and Google Cloud Storage connector (https://cloud.google.com/dataproc/docs/concepts/connectors/cloud-storage).
Google Cloud Storage connector (GCS connector) has a dependency problem with Apache Spark releated with Google Guava, so I added this library's shaded jar in server/lib (SBT Unmanaged dependencies, not Managed dependencies in build.sbt).
I implemented this very forcefully, is this no problem?

To use GCP for delta sharing server, you have to do 2 steps.

1. In your terminal, Do `export GOOGLE_APPLICATION_CREDENTIALS=</path/to/your_gcp_service_acount_key.json>`
2. In the server config (e,g, delta-sharing-server.yaml), add a setting for GCP
```
shares:
- name: "share_gcp"
  schemas:
  - name: "schema_gcp"
    tables:
    - name: "table_gcp"
      location: "gs://delta-start/delta/"
```

If you want to coneect this server from Apache Spark,  execute spark-shell or spark-submit like below.

```
spark-shell \
--packages io.delta:delta-core_2.12:1.0.0,io.delta:delta-contribs_2.12:1.0.0,io.delta:delta-sharing-spark_2.12:0.2.0 \
--jars </path/to/gcs-connector-latest-hadoop2.jar> \
--conf spark.delta.logStore.oci.impl=io.delta.storage.GCSLogStore \
--conf spark.hadoop.fs.gs.impl=com.google.cloud.hadoop.fs.gcs.GoogleHadoopFS \
--conf spark.hadoop.google.cloud.auth.service.account.enable=true 
--conf spark.hadoop.google.cloud.auth.service.account.json.keyfile=</path/to/your_gcp_service_acount_key.json>
```

I haven't implement integration test on GCP support yet.
If I understand correctly, what I should do is to add the almost same test with Azure support (https://github.com/delta-io/delta-sharing/blob/main/server/src/test/scala/io/delta/sharing/server/DeltaSharingServiceSuite.scala#L510).
Is this correct?
If my understanding is correct, I'll add this integreation test later.

### CODE REVIEW

**Comment by zsxwing** [comment]
Thanks for the contribution. This is awesome!

> so I added this library's shaded jar in server/lib (SBT Unmanaged dependencies, not Managed dependencies in build.sbt).

Is Guava the only issue? If so, we can just exclude guava from other dependencies like this https://github.com/delta-io/delta-sharing/commit/a6bd550fd7248f8e667babeb21f946539d2c0ec3 This is better than adding a shaded jar to the project.

> In your terminal, Do `export GOOGLE_APPLICATION_CREDENTIALS=</path/to/your_gcp_service_acount_key.json>`

Does the GOOGLE_APPLICATION_CREDENTIALS env work for reading files using `GoogleHadoopFileSystem`? I saw GCS connector asks users to set a hadoop configuration `google.cloud.auth.service.account.json.keyfile`: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/INSTALL.md#configuring-hadoop Since this is the recommended way, can we just read the key file from the hadoop configuration? This will be similar to https://github.com/delta-io/delta-sharing#azure-data-lake-storage-gen2

> If I understand correctly, what I should do is to add the almost same test with Azure support 

Correct. I will do some manual test with your change for now. We can add the real integration tests later.

**Comment by zsxwing** [comment]
@kohei-tosshy Any thoughts on my above questions?

**Comment by kohei-tosshy** [comment]
Thank you for your comment.
And sorry for being late for reacting.

> Is Guava the only issue? If so, we can just exclude guava from other dependencies like this a6bd550 This is better than adding a shaded jar to the project.

I see. I'll try.

> Does the GOOGLE_APPLICATION_CREDENTIALS env work for reading files using GoogleHadoopFileSystem? I saw GCS connector asks users to set a hadoop configuration google.cloud.auth.service.account.json.keyfile: https://github.com/GoogleCloudDataproc/hadoop-connectors/blob/master/gcs/INSTALL.md#configuring-hadoop Since this is the recommended way, can we just read the key file from the hadoop configuration? This will be similar to https://github.com/delta-io/delta-sharing#azure-data-lake-storage-gen2

To get a presigned URL, I used Google Google Cloud Client Library.
I only used `GoogleHadoopFileSystem` only to changing a path.
GOOGLE_APPLICATION_CREDENTIALS is needed to use Google Google Cloud Client Library.

> Correct. I will do some manual test with your change for now. We can add the real integration tests later.

I see. I'll try.

**Comment by kohei-tosshy** [comment]
@zsxwing

Sorry for being late for revising.
I revised Google Guava's dependency problem and added an integration test for GCS support.

File: .gitignore [hunk 1]
```diff
@@ -21,7 +21,6 @@
 .pydevproject
 .scala_dependencies
 .settings
-/lib/
```
Comment (zhuansunxt) [inline .gitignore:24]: Revert this change?

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala [hunk 2]
```diff
@@ -188,3 +193,20 @@ object AbfsFileSigner {
...
+
+class GCSFileSigner(
+    name: URI,
+    conf: Configuration,
+    preSignedUrlTimeoutSeconds: Long) extends CloudFileSigner{
+
+  override def sign(path: Path): String = {
+    val absPath = path.toUri
+    val bucketName = absPath.getHost
+    val objectName = absPath.getPath.stripPrefix("/")
```
Comment (zhuansunxt) [inline server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala:205]: Would be great to have a check `assert(objectName.nonEmpty, s"cannot get object key from $path")` similar to AWS and Azure signer

**Review by zhuansunxt — APPROVED** [review]
The code change LGTM.

@zsxwing Do you want to have another look? It'll be nice to include this in the upcoming release.

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala [hunk 3]
```diff
@@ -188,3 +193,20 @@ object AbfsFileSigner {
       getRelativePath(abfsStore, _))
   }
 }
+
+class GCSFileSigner(
+    name: URI,
+    conf: Configuration,
+    preSignedUrlTimeoutSeconds: Long) extends CloudFileSigner{
```
Comment (zsxwing) [inline server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala:200]: nit
```suggestion
    preSignedUrlTimeoutSeconds: Long) extends CloudFileSigner {
```

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala [hunk 4]
```diff
@@ -188,3 +193,20 @@ object AbfsFileSigner {
...
+
+  override def sign(path: Path): String = {
+    val absPath = path.toUri
+    val bucketName = absPath.getHost
+    val objectName = absPath.getPath.stripPrefix("/")
+    val storage = StorageOptions.newBuilder.build.getService
```
Comment (zsxwing) [inline server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala:206]: Is `storage` thread-safe? If so, we can save it as a val in `GCSFileSigner` to avoid loading the credentials for each url.

File: server/src/test/scala/io/delta/sharing/server/DeltaSharingServiceSuite.scala [hunk 5]
```diff
@@ -551,4 +551,34 @@ class DeltaSharingServiceSuite extends FunSuite with BeforeAndAfterAll {
     }
     assert(e.getMessage.contains("Server returned HTTP response code: 403")) // 403 Forbidden
   }
+
+  integrationTest("gcs support") {
```
Comment (zsxwing) [inline server/src/test/scala/io/delta/sharing/server/DeltaSharingServiceSuite.scala:555]: nit: Could you use `ignore` for this test for now? We will set up the credentials and enable it later.

**Comment by zsxwing** [comment]
> To get a presigned URL, I used Google Google Cloud Client Library.
I only used GoogleHadoopFileSystem only to changing a path.
GOOGLE_APPLICATION_CREDENTIALS is needed to use Google Google Cloud Client Library.

Google Google Cloud Client Library also supports other ways to config credentials. But it seems hard to mimic what GoogleHadoopFileSystem does to support various of ways to config credentials. `GOOGLE_APPLICATION_CREDENTIALS` env is a good start right now. We can support other ways in future if people are asking.

Left some minor comments. Otherwise LGTM.

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala (hunk 2 above)
Comment (kohei-tosshy) [reply server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala:207]: I added it.

File: .gitignore (hunk 1 above)
Comment (kohei-tosshy) [reply .gitignore]: I reverted it.

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala (hunk 3 above)
Comment (kohei-tosshy) [reply server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala]: I revised this.

File: server/src/test/scala/io/delta/sharing/server/DeltaSharingServiceSuite.scala (hunk 5 above)
Comment (kohei-tosshy) [reply server/src/test/scala/io/delta/sharing/server/DeltaSharingServiceSuite.scala]: I used `ignore`  instead of `integrationTest`.

File: server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala (hunk 4 above)
Comment (kohei-tosshy) [reply server/src/main/scala/io/delta/sharing/server/CloudFileSigner.scala]: I think `storage` is thread-safe.
One instance of `GCSFileSigner` is created in each request for client, so each thread for a request has a dedicated instance of `GCSFileSigner` and won't touch other `GCSFileSigner` instances.
And, `storage` is used only for reading, so it will be OK if some threads for creating a presignied-URL are processed in paralle.

**Review by zsxwing — APPROVED** [review]
LGTM! Thanks!
